### PR TITLE
feat(sponnet): extend pipeline library

### PR DESCRIPTION
**Pull request by kskewes**
RE: #3456 

Add jenkins stage
Add artifact option for deploy manifest
Add notifications
Add stage timeout
Add moniker
Add expectedArtifacts
Add triggers (docker and git only) constraints TODO

Notes:

The `jsonnet fmt` function is responsible for a most of the changes to pre-existing lines.
Inheritance using a sleek per app `appname.jsonnet` is not catered to here and subject to outcomes from the design doc on MPT v2, future PR's, etc.

Questions:

1. camelCase for variables and functions is common in other projects like Prometheus, Databricks. I'm tempted to change `deploy_manifest` etc. Thoughts?

### CODE REVIEW

File: sponnet/pipeline.libsonnet [hunk 1]
```diff
@@ -2,52 +2,118 @@
...
     id: id,
-    withMatchArtifact(matchArtifact):: self + {matchArtifact: matchArtifact},
-    withDefaultArtifact(defaultArtifact):: self + {defaultArtifact: defaultArtifact},
-    withUsePriorArtifact(usePriorArtifact):: self + {usePriorArtifact: usePriorArtifact},
-    withUseDefaultArtifact(useDefaultArtifact):: self + {useDefaultArtifact: useDefaultArtifact},
+    // Defaults to kind: custom
+    defaultArtifact: { kind: 'custom' },
+    withMatchArtifact(matchArtifact):: self + {
+      matchArtifact+: {
+        kind: std.splitLimit(matchArtifact.type, '/', 1)[0],
```
Comment (lwander) [inline sponnet/pipeline.libsonnet]: Kind is just for the UI, and doesn't (as far as I know) always follow this pattern. I'd drop this.

File: sponnet/pipeline.libsonnet [hunk 2]
```diff
@@ -2,52 +2,118 @@
...
-    githubFile():: artifact("github/file"),
-    gcsObject():: artifact("gcs/object"),
+    dockerImage():: artifact('docker/image'),
+    githubFile():: artifact('github/file'),
+    gitlabFile():: artifact('gitlab/file'),
+    gcsObject():: artifact('gcs/object'),
   },
 
+  // expected artifacts
+  // TODO: This section may need splitting out by artifact type due to differing field requirements.
```
Comment (lwander) [inline sponnet/pipeline.libsonnet]: :+1:

File: sponnet/pipeline.libsonnet (hunk 1 above)
Comment (kskewes) [reply sponnet/pipeline.libsonnet]: Dropped.

Comment (kskewes) [reply sponnet/pipeline.libsonnet]: https://github.com/spinnaker/spinnaker/issues/3491

**Comment by lwander** [comment]
Nice work!

**Comment by lwander** [comment]
Also, :+1: to changing from `deploy_manifest` to `deployManifest`, given that we are camel case in most of the sponnet library anyhow. Let's do that in a separate PR

**Pull request merged**